internal/dag/policy.go

```diff
@@ -57,6 +57,9 @@ const (
 	LoadBalancerPolicyRequestHash = "RequestHash"
 )
 
+// match "%REQ(<X-Foo-Bar>)%"
+var hostRewriteHeaderRegex = regexp.MustCompile(`%REQ\((\S+)\)%`)
```

Comment:
In access log validation we do have regex/validation for complex access log operators like this: https://github.com/projectcontour/contour/blob/5b564bc97cd37b57710225e3d3cdbefeba16293a/apis/projectcontour/v1alpha1/accesslog.go#L282-L292

This does also support truncation of the value the operator returns

Not sure if we want to support truncation or the host rewrite from header actually supports that but, could be some ability to consolidate code

Reply:
Also in this case do we need beginning of string and end of line anchors for this regex? Can you do something where you set the host header to a value like this: `foo.%REQ(X-Other-Header).com`? or is that illegal?

Reply:
The host rewrite via header(this PR) or via literal(already on main)are separate fields on the envoy config. 
Currently, on main, if the header is set to `host`, we use the value as the host rewrite literal. To set the host via dynamic header, I need some signal other than the `host` header to set the host_rewrite_header field. 

This is why I decided to use `%REQ(X-Other-Header)%`, and in Contour, I strip out the irrelevant bits and only use `X-Other-Header` in the envoy config. So it's not an envoy thing rather I wanted to keep the contour surface consistent. 

As for the envoy host_rewrite_header itself, I think it needs a header name and the format `%DOWNSTREAM_LOCAL_ADDRESS% `doesn't seem to work. 

example: 
envoy config bits: 
```
{
        "name": "local.projectcontour.io",
        "domains": [
         "local.projectcontour.io"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/"
          },
          "route": {
           "cluster": "echo/ingress-conformance-echo/80/da39a3ee5e",
           "host_rewrite_header": "%DOWNSTREAM_LOCAL_ADDRESS%"
          },
          "metadata": {
           "filter_metadata": {
            "envoy.access_loggers.file": {
             "io.projectcontour.kind": "HTTPProxy",
             "io.projectcontour.name": "echo",
             "io.projectcontour.namespace": "echo"
            }
           }
          },
          "request_headers_to_add": [
           {
            "header": {
             "key": "Test",
             "value": "%DOWNSTREAM_LOCAL_ADDRESS%"
            },
            "append_action": "OVERWRITE_IF_EXISTS_OR_ADD"
           }
          ]
         }
        ]
       }
```

```
curl --location --request GET 'local.projectcontour.io'
{"TestId":"","Path":"/","Host":"local.projectcontour.io","Method":"GET","Proto":"HTTP/1.1","Headers":{"Accept":["*/*"],"Test":["10.244.0.2:8080"],"User-Agent":["curl/8.1.2"],"X-Envoy-Expected-Rq-Timeout-Ms":["15000"],"X-Envoy-Internal":["true"],"X-Forwarded-For":["192.168.65.1"],"X-Forwarded-Proto":["http"],"X-Request-Id":["b0c0bbf7-b204-4edc-876c-3926a14141b3"],"X-Request-Start":["t=1695284772.535"]}}%
```

Reply:
postive case 

```
 {
        "name": "local.projectcontour.io",
        "domains": [
         "local.projectcontour.io"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/"
          },
          "route": {
           "cluster": "echo/ingress-conformance-echo/80/da39a3ee5e",
           "host_rewrite_header": "X-Header-Name"
          },
          "metadata": {
           "filter_metadata": {
            "envoy.access_loggers.file": {
             "io.projectcontour.name": "echo",
             "io.projectcontour.namespace": "echo",
             "io.projectcontour.kind": "HTTPProxy"
            }
           }
          },
          "request_headers_to_add": [
           {
            "header": {
             "key": "Test",
             "value": "%DOWNSTREAM_LOCAL_ADDRESS%"
            },
            "append_action": "OVERWRITE_IF_EXISTS_OR_ADD"
           }
          ]
         }
        ]
       }
```

```
curl --location --request GET 'local.projectcontour.io' \
--header 'x-header-name: public.projectcontour.io'
{"TestId":"","Path":"/","Host":"public.projectcontour.io","Method":"GET","Proto":"HTTP/1.1","Headers":{"Accept":["*/*"],"Test":["10.244.0.2:8080"],"User-Agent":["curl/8.1.2"],"X-Envoy-Expected-Rq-Timeout-Ms":["15000"],"X-Envoy-Internal":["true"],"X-Forwarded-For":["192.168.65.1"],"X-Forwarded-Proto":["http"],"X-Header-Name":["public.projectcontour.io"],"X-Request-Id":["5152a296-22c3-4b9a-9c94-a26fc7456673"],"X-Request-Start":["t=1695284874.873"]}}%
```

Reply:
Ah i see, i didnt realize the Envoy field just took a header name, so we just need a sentinel value to let us know we want to rewrite the host with the content of a header